### ISSUE TITLE
kafka slice taking significantly longer than wait time

### ISSUE THREAD

**Issue by ciorg** [terascope/kafka-assets#11]
While running a job I noticed that the kafka_reader was taking 30 min to complete a slice instead of the specified wait time of 30 sec.
Related job details:
```
 {
            "_op": "teraslice_kafka_reader",
             ...
            "size": 100000,
            "wait": 30000
}
```
I reduced the size to 25,000 and now the kafka_reader takes about 6 minutes per slice.

This only happens when the job is waiting for data, but is consistent with the jobs that read from this particular topic.  Others job reading a different topic show the kakfa_reader finishing slices every 30 sec whether there is data or not.  

Both topics are basically json, but the dataset with the long slice times has many more fields and is more complex.

**Comment by macgyver603** [comment]
This shows up specifically when a consumer is reading from a topic with a constant, low-volume data stream. In my test, I started a 3-worker Teraslice job with the same `size` and `wait` specified in your comment that was reading from a freshly-created topic with no records or producers, and after letting it sit for a short period on it's own, it was able to complete about 15 empty slices with an average execution time of 30s for the `teraslice_kafka_reader`, showing that the `wait` parameter was respected initially. Then, I started up a Teraslice data generator with a single worker that was capable of producing no more than 700 records/sec, and the average execution time of the `teraslice_kafka_reader` op on my original job shot up to ~7min. To take it a step further after a couple of 7min completion times, I kicked off a `once` data generator with 15 workers just to add 10mil records to the topic as quickly as possible. It ended up adding a bit more lag than I had wanted, but while the reader job caught up, the average execution time for the `teraslice_kafka_reader` was averaging ~4.5s to complete, but after it caught back up on the lag, the average execution time jumped back up to 7min.

My theory for the root cause of this is that the way the consumer is implemented, `consumer.consume()` will not return until EITHER the producers stop writing to the topic long enough for a consume request to finish OR kafka is done returning the full number of records. This would explain why jobs reading from low-volume but bursty topics will respect the `wait` parameter between the bursts, but jobs reading from low-volume topics with a constant trickle of records will just continue collecting records until it has the full chunk of data, seemingly ignoring the `wait` parameter. 

Looking at [`consume()`](https://github.com/terascope/kafka-assets/blob/v1.2.0/asset/teraslice_kafka_reader/index.js#L202-L216), the processor looks at the `wait` parameter only if `readyToProcess ` is true. However, once the a consume is started, `readyToProcess` is toggled to false, blocking any further attempts to consume or resolve the slice, and it is not toggled back to true until the consume finishes and the worker is done adding the records to its current slice.

A really interesting thing to note is that I ran this test in a Kubernetes Teraslice cluster, so I was able to see worker-level metrics. From the network IO chart (averaged across the three workers), it looks like the records are being sent to the workers as they become available in the topic instead of being dumped on the worker all at once.
![screen shot 2018-11-16 at 4 15 08 pm](https://user-images.githubusercontent.com/11929901/48652125-242ec080-e9bb-11e8-9f66-c39d3a35610a.png)
